### PR TITLE
Feature/reservations

### PR DESCRIPTION
**Pull request by chGoodchild**
Added OCPP reservations without any caching of reservations in the OCPP charge point:
* Switched from `std::chrono::system_clock()` to `date::utc_clock()` to avoid discrepancies with absolute time points.
* Reservations are only stored by EVerest.

### CODE REVIEW

File: modules/EvseManager/EvseManager.cpp [hunk 1]
```diff
@@ -75,6 +104,60 @@ json EvseManager::get_hw_capabilities() {
...
+
+    // is it already reserved with a different reservation_id?
+    if (reservation_valid() && _reservation_id != reservation_id)
+        return "Rejected";
+
+    std::lock_guard<std::mutex> lock(reservation_mutex);
+
+    // accept new reservation
+    reserved_auth_token = token;
+    EVLOG(critical) << "valid_until: " << valid_until;
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:135]: Remove critical log messages

File: modules/EvseManager/EvseManager.cpp [hunk 2]
```diff
@@ -86,8 +169,97 @@ bool EvseManager::updateLocalMaxCurrentLimit(float max_current) {
...
+
+// Signals that reservation was used to start this charging.
+// Does nothing if no reservation is active.
+void EvseManager::use_reservation_to_start_charging() {
+    #ifndef NDEBUG
+        EVLOG(critical) << "Reserved: " << reserved;
+    #endif
+
+    std::lock_guard<std::mutex> lock(reservation_mutex);
+    if (!reserved)
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:208]: if block { }

File: modules/EvseManager/EvseManager.cpp [hunk 3]
```diff
@@ -86,8 +169,97 @@ bool EvseManager::updateLocalMaxCurrentLimit(float max_current) {
     return false;
 }
 
+bool EvseManager::cancel_reservation() {
+    bool res_valid = reservation_valid();
+    #ifndef NDEBUG
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:174]: Could these messages become "debug" messages?

File: modules/EvseManager/EvseManager.cpp [hunk 4]
```diff
@@ -86,8 +169,97 @@ bool EvseManager::updateLocalMaxCurrentLimit(float max_current) {
...
+    return false;
+}
+
+/*
+  Returns false if not reserved or reservation matches the token.
+  Returns true if reserved for a different token.
+*/
+bool EvseManager::reserved_for_different_token(const std::string& token) {
+    std::lock_guard<std::mutex> lock(reservation_mutex);
+    if (!reserved)
```
Comment (hikinggrass) [inline modules/EvseManager/EvseManager.cpp:257]: if block { }

File: modules/OCPP/OCPP.hpp [hunk 5]
```diff
@@ -57,7 +59,19 @@ class OCPP : public Everest::ModuleBase {
     const std::unique_ptr<ocpp_1_6_charge_pointImplBase> p_main;
     const std::unique_ptr<auth_token_validatorImplBase> p_auth_validator;
     const std::vector<std::unique_ptr<evse_managerIntf>> r_evse_manager;
-
+    std::map<int32_t, int32_t> ResConnMap;
```
Comment (hikinggrass) [inline modules/OCPP/OCPP.hpp:62]: I think we should continue using snake case for consistency here. eg. res_conn_map and res_state_map. Maybe drop the _map postfix as well

File: modules/YetiDriver/yeti_comms/CMakeLists.txt [hunk 6]
```diff
@@ -17,4 +17,8 @@ add_library(yeti_comms STATIC evSerial.cpp ../common/evThread.cpp protobuf/hi2lo
 target_include_directories(yeti_comms PUBLIC "${PROJECT_BINARY_DIR}" PUBLIC "nanopb" PUBLIC "protobuf" PUBLIC "../common")
 target_link_libraries(yeti_comms PRIVATE Pal::Sigslot Threads::Threads)
 
-
+target_link_libraries(yeti_comms
+    PRIVATE
+        date::date
+        date::date-tz
+)
```
Comment (hikinggrass) [inline modules/YetiDriver/yeti_comms/CMakeLists.txt:24]: Missing newline at end of file

File: run_sil-ocpp-complete-fixture.sh [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
```
Comment (hikinggrass) [inline run_sil-ocpp-complete-fixture.sh:1]: Remove, seems redundant
How is this file different from run_sil-ocpp-complete.sh other than using another directory level?

File: tests/test_reserve_now.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
+from paho.mqtt import client as mqtt_client
```
Comment (hikinggrass) [inline tests/test_reserve_now.py:1]: I'm not sure about including this just yet, things especially with python based testing will change quite a bit in the near future. I would rather remove this for now and introduce it when all the pieces are in place

**Review by hikinggrass — CHANGES_REQUESTED** [review]

**Comment by chGoodchild** [comment]
It runs again, but I have unsigned commits before Kai's commit: fc52f57162cc8045d0aa4ae73ce43a081ecbd183
I will fix this by: 
* resetting the head to this commit: fced91f8104ec8910aeabcb37f55c45a590b42e7
* signing my commits
* creating a new branch with the signed commits
* cherry picking Kai's commits into my new branch
* cherry picking my latest commits into the new branch
* signing my latest commits
* and overwriting this branch

Then it should be ready to merge :)

**Review by hikinggrass — APPROVED** [review]

**Review by hikinggrass — APPROVED** [review]

**Pull request merged**